From: Kefu Chai <kchai@redhat.com>
Date: Wed, 23 Aug 2017 16:34:12 +0800
Subject: osd/PGBackend: delete reply if fails to complete delete request

if any of the objects fails to be deleted due to pg reset after latest
osdmap, the pg recovery delete reply won't be sent to the primary OSD.
in that case, we should delete the reply.

Fixes: http://tracker.ceph.com/issues/20913
Signed-off-by: Kefu Chai <kchai@redhat.com>
(cherry picked from commit 1c18b5cb0c27d7976e6d3d5e4ea6c3935685019b)

Resolves: rhbz#1452780
---
 src/osd/PGBackend.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/osd/PGBackend.cc b/src/osd/PGBackend.cc
index f634d20..df27e83 100644
--- a/src/osd/PGBackend.cc
+++ b/src/osd/PGBackend.cc
@@ -133,6 +133,8 @@ void PGBackend::handle_recovery_delete(OpRequestRef op)
     [=](int r) {
       if (r != -EAGAIN) {
 	get_parent()->send_message_osd_cluster(reply, conn.get());
+      } else {
+	delete reply;
       }
     }));
   gather.activate();
